codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java

```diff
@@ -506,66 +392,60 @@ private void insertHeader(Http2Headers headers, CharSequence name, CharSequence
-            headerSize = (int) newSize;
-        } else {
-            // truncation will be reported during endHeaderBlock
-            headerSize = maxHeaderSize + 1;
-        }
+        return headerSize;
     }
 
-    private boolean exceedsMaxHeaderSize(long size) {
-        // Check new header size against max header size
-        if (size + headerSize <= maxHeaderSize) {
-            return false;
+    private int addHeader(Http2Headers headers, CharSequence name, CharSequence value, int headersLength)
+            throws Http2Exception {
+        long newHeadersLength = (long) headersLength + name.length() + value.length();
+        if (newHeadersLength > maxHeadersLength) {
+            maxHeaderSizeExceeded();
         }
+        headers.add(name, value);
+        return (int) newHeadersLength;
+    }
 
-        // truncation will be reported during endHeaderBlock
-        headerSize = maxHeaderSize + 1;
-        return true;
+    /**
+     * Respond to headers block resulting in the maximum header size being exceeded.
+     * @throws Http2Exception If we can not recover from the truncation.
+     */
+    private void maxHeaderSizeExceeded() throws Http2Exception {
+        throw connectionError(ENHANCE_YOUR_CALM, "Header size exceeded max allowed bytes (%d)", maxHeadersLength);
     }
 
-    private CharSequence readStringLiteral(ByteBuf in, int length) throws IOException {
+    private CharSequence readStringLiteral(ByteBuf in, int length, boolean huffmanEncoded) throws Http2Exception {
         if (huffmanEncoded) {
             return huffmanDecoder.decode(in, length);
-        } else {
-            byte[] buf = new byte[length];
-            in.readBytes(buf);
-            return new AsciiString(buf, false);
         }
+        byte[] buf = new byte[length];
+        in.readBytes(buf);
+        return new AsciiString(buf, false);
+    }
+
+    private static IllegalArgumentException notEnoughDataException(ByteBuf in) {
+        return new IllegalArgumentException("decode only works with an entire header block! " + in);
     }
 
     // Unsigned Little Endian Base 128 Variable-Length Integer Encoding
-    private static int decodeULE128(ByteBuf in) throws IOException {
-        in.markReaderIndex();
-        int result = 0;
-        int shift = 0;
-        while (shift < 32) {
-            if (!in.isReadable()) {
-                // Buffer does not contain entire integer,
-                // reset reader index and return -1.
-                in.resetReaderIndex();
-                return -1;
-            }
-            byte b = in.readByte();
+    private static int decodeULE128(ByteBuf in) throws Http2Exception {
+        int readerIndex = in.readerIndex();
+        final int writerIndex = in.writerIndex();
+        for (int shift = 0, result = 0; readerIndex < writerIndex; ++readerIndex, shift += 7) {
+            byte b = in.getByte(readerIndex);
             if (shift == 28 && (b & 0xF8) != 0) {
```

Comment:
If you look through the spec, the smallest partial byte is 4 bits (Ctrl F for `+)` in the spec).  Per additional byte, the number of bits is:
```
4  - 0 bytes
11 - 1 byte
18 - 2 bytes
25 - 3 bytes
32 - 4 bytes
```

Since Java ints are signed, and I don't think its possible to have a legitimate 32 bits would be too many.  Since you are only looking at the *extra* bytes after the first partial one, I think that getting a 4th byte would mean overflow.

Here, it check if the shift value is 28, but by that point its already too much.  We know for sure smallest possible partial byte would cause an overflow, right?